core/src/main/java/io/questdb/cutlass/text/TextImportTask.java

```diff
@@ -0,0 +1,992 @@
+                utf8Sink.clear();
+                fd = TableUtils.openRO(ff, path, LOG);
+
+                long MASK = ~((255L) << 56 | (255L) << 48);
+                final long count = size / (2 * Long.BYTES);
+                long bytesToRead;
+                int additionalLines;
+
+                for (long i = 0; i < count; i++) {
+                    long lengthAndOffset = Unsafe.getUnsafe().getLong(address + i * 2L * Long.BYTES + Long.BYTES);
+                    long lineLength = lengthAndOffset >>> 48;
+                    offset = lengthAndOffset & MASK;
+                    bytesToRead = lineLength;
+                    additionalLines = 0;
+
+                    for (long j = i + 1; j < count; j++) {
+                        long nextLengthAndOffset = Unsafe.getUnsafe().getLong(address + j * 2L * Long.BYTES + Long.BYTES);
+                        long nextLineLength = nextLengthAndOffset >>> 48;
+                        long nextOffset = nextLengthAndOffset & MASK;
+
+                        //line indexing stops on first EOL char, e.g. \r, but it could be followed by \n
+                        long diff = nextOffset - offset - bytesToRead;
+                        if (diff > -1 && diff < 2 && bytesToRead + nextLineLength < fileBufSize) {
+                            bytesToRead += diff + nextLineLength;
+                            additionalLines++;
+                        } else {
+                            break;
+                        }
+                    }
+
+                    i += additionalLines;
+
+                    long n = ff.read(fd, fileBufAddr, bytesToRead, offset);
+                    if (n > 0) {
+                        lexer.parse(fileBufAddr, fileBufAddr + n, Integer.MAX_VALUE, this::onFieldsPartitioned);
+                    } else {
+                        throw TextException.$("Can't read from file path='").put(path).put("', errno=").put(ff.errno()).put(",offset=").put(offset);
+                    }
+                }
+
+                lexer.parseLast();
+            } finally {
+                if (fd > -1) {
+                    ff.close(fd);
+                }
+            }
+        }
+
+        private void mergePartitionIndexAndImportData(
+                final FilesFacade ff,
+                Path partitionPath,
+                final TextLexer lexer,
+                long fileBufAddr,
+                long fileBufSize,
+                DirectCharSink utf8Sink,
+                DirectLongList mergeIndexes,
+                Path tmpPath
+        ) throws TextException {
+            mergeIndexes.clear();
+            partitionPath.slash$();
+            int partitionLen = partitionPath.length();
+
+            long mergedIndexSize = -1;
+            long address = -1;
+            long fd = -1;
+
+            try {
+                mergedIndexSize = openIndexChunks(ff, partitionPath, mergeIndexes, partitionLen);
```

Comment:
handling of single index (e.g. no merge necessary) should be separate execution path, e.g. without conditional unmaps and conditional pass-thru calls. Perhaps handle this in Java rather than C

Reply:
Fixed